content/blog/2025-04-24-super-optimization/index.md

```diff
@@ -0,0 +1,104 @@
+
+## Scaling to different architectures and optimization targets
+
+For some optimization targets, such as code size, it might be straightforward to come up with a precise cost model, while for targets like power consumption, it might be impossibly hard. When porting existing Denali frameworks to other architectures, many of the mathematical axioms can be reused, but extra human efforts for rewriting the architectural axioms are still required. 
+
+If we take the approach of optimizing on the IR level, we can potentially make our cost model more scalable across different architectures and efficient given the reduced search space since most of the architectural details are abstracted away. But that gain might come at the price of being more imprecise. Maybe superoptimization is in some sense not scalable by definition. We really need architecture specific information to find the optimal solution for a given target. 
+
+## Other related work	
+	
+**Superoptimizer -- A Look at the Smallest Program (1987)**
+
+
+Massalin’s original superoptimizer brute forces all enumerations of machine code sequence in a BFS manner, and accepts the shortest program that generates the correct result for all test suites. Correctness of the output program cannot be guaranteed, even if all prepared test suites are passed. The huge search space limits the length of the synthesized code sequence and the number of opcodes to explore. This approach only succeeded in finding impressive short code sequences with length around half-a-dozen instructions. With the restriction of limiting the opcodes to consider between 10-15, it can find solutions of length up to a dozen instructions. 
+
+**STOKE: Stochastic Superoptimization (2013)**
+
+
+The STOKE superoptimizer formulates the superoptimization task as a stochastic search problem and search with MCMC. The cost function is defined as a combination of correctness and perf improvement. This approach trades off between completeness and the size of search space. Some human annotations are required to help generate random program input used to approximate the cost function value. The level of human intervention is considered to be light work compared with the other two approaches we mentioned and expert knowledge is not required. This approach is able to handle CISC ISA (x86_64) with over 400 opcodes.
+
+
+**Souper: A Synthesizing Superoptimizer (2018)**
+
+Souper takes a different approach to the other superoptimizers discussed, in that instead of outputting assembly, its output is IR: it both takes in, and produces as output, LLVM IR. Its algorithm is CEGIS: enumeration and testing, like Massalin’s superoptimizer, but in a loop with a verification tool that can add tests. It’s designed to be used mainly by compiler engineers, to help improve compiler optimizations: both LLVM and MSVC include optimizations inspired by suggestions from the output of Souper.
```

Comment:
> CEGIS: enumeration and testing, like Massalin’s superoptimizer, but in a loop with a verification tool that can add tests

That seems like _almost_ a full description of (my understanding of CEGIS)… I usually think of the CEGIS loop relying on an SMT solver (not simply enumeration) to make proposals. But it's been a while since I read the Souper paper so I'm not sure that's true of their approach…